### PR TITLE
Support negative pad values

### PR DESCRIPTION
**Pull request by codeislife99**
Negative pad values error out currently. This PR removes it and handles the error correctly for negative pad values.

### CODE REVIEW

**Comment by codeislife99** [comment]
@anijain2305 @comaniac @zhiics @trevor-m PTAL

**Comment by anijain2305** [comment]
Please add a unit test

**Review by trevor-m — CHANGES_REQUESTED** [review]
Thanks @codeislife99 ! It looks good to me, but could you also add a test case in `test_op_level2.py`?

**Comment by codeislife99** [comment]
Added relevant test cases.

**Comment by codeislife99** [comment]
@anijain2305 @trevor-m 
One dilemna here is that when shape is dynamic I cannot check if the existing shape + padding would be greater than zero in infertype. This will lead to a memory error during executing when it wont be able to create memory for negative tensor sizes. And the error would be quite bad because the developer/user would not have a clue where it is coming from in a big model. 

Does anyone have any idea of how to solve this ? Should we display a warning whenever we encounter negative padding ?

File: tests/python/relay/test_op_level2.py [hunk 1]
```diff
@@ -1171,35 +1171,64 @@ def test_flatten_infer_type():
...
     t = relay.var("t", relay.TensorType((n, c, h, w), "float32"))
     y = relay.nn.pad(t, ((1, 1), (2, 2), (3, 3), (4, 4)))
     "pad_width=" in y.astext()
     yy = run_infer_type(y)
     assert yy.checked_type == relay.TensorType((3, 6, 9, 12), "float32")
 
+    n, c, h, w = 4, 6, 3, 5
+    t = relay.var("t", relay.TensorType((n, c, h, w), "float32"))
+    y = relay.nn.pad(t, ((-1, -1), (2, -2), (0, -3), (4, 4)), pad_mode="reflect")
+    "pad_width=" in y.astext()
```
Comment (comaniac) [inline tests/python/relay/test_op_level2.py:1185]: Is that what you meant?
```suggestion
    assert "pad_width=" in y.astext()
```

**Review by comaniac — COMMENTED** [review]
When handling the dynamic shape, you need to improve the shape function to check if the shape after padding is still legal. If it cannot be simply covered in this PR, we can focus on the static shape first and still throw out errors for negative padding values when the shape is dynamic.

File: tests/python/relay/test_op_level2.py (hunk 1 above)
Comment (codeislife99) [reply tests/python/relay/test_op_level2.py:1185]: Oops, this is an artifact by copying over from the previous example.

**Review by comaniac — APPROVED** [review]
LGTM

File: tests/python/relay/test_op_level2.py [hunk 2]
```diff
@@ -1171,35 +1171,62 @@ def test_flatten_infer_type():
...
-        x = relay.var("x", shape=dshape)
-        y = relay.nn.pad(x, ((1, 1), (2, 2), (3, 3), (4, 4)))
-        func = relay.Function([x], y)
-        data = np.random.uniform(size=dshape).astype(dtype)
-        ref_res = np.pad(data, ((1, 1), (2, 2), (3, 3), (4, 4)), "constant")
-        for target, ctx in tvm.testing.enabled_targets():
-            intrp1 = relay.create_executor("graph", ctx=ctx, target=target)
-            op_res1 = intrp1.evaluate(func)(data)
-            tvm.testing.assert_allclose(op_res1.asnumpy(), ref_res, rtol=1e-5, atol=1e-5)
+        dshape_list = [(4, 10, 7, 7), (4, 6, 3, 5)]
```
Comment (trevor-m) [inline tests/python/relay/test_op_level2.py:1204]: It seems a bit odd to use a test case which results in an empty tensor (dimension of 0). It means there won't be any actual values to check if the result is correct. Could you change the either the dim with value of 3 or the `-3` padding?

**Review by trevor-m — CHANGES_REQUESTED** [review]

File: tests/python/relay/test_op_level2.py (hunk 2 above)
Comment (codeislife99) [reply tests/python/relay/test_op_level2.py:1204]: Thats a good point, resolved !

**Review by trevor-m — APPROVED** [review]

**Review by anijain2305 — APPROVED** [review]
LGTM

**Comment by anijain2305** [comment]
Thanks @codeislife99 @comaniac @trevor-m This is merged

**Pull request merged**